### PR TITLE
refactor: rework vector abstraction

### PR DESCRIPTION
**Pull request by usamoi**
We will finish most of tasks in #255 after this PR is merged.

### CODE REVIEW

File: src/datatype/binary_svecf32.rs
Comment (cutecutecat) [inline src/datatype/binary_svecf32.rs:1]: Why don't use structure like:
- svecf32/binary.rs
- svecf32/text.rs
- svecf32/operator.rs
- svecf32/subscript.rs
- svecf32/memory.rs

File: crates/base/src/global/svecf32_cos.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    const VECTOR_KIND: VectorKind = VectorKind::SVecf32;
+    const DISTANCE_KIND: DistanceKind = DistanceKind::Cos;
+
+    fn distance(lhs: Borrowed<'_, Self>, rhs: Borrowed<'_, Self>) -> F32 {
+        F32(1.0) - super::svecf32::cosine(lhs, rhs)
+    }
+}
+
+impl GlobalElkanKMeans for SVecf32Cos {
+    fn elkan_k_means_normalize(vector: &mut [Scalar<Self>]) {
```
[the remark this answers is not in the document]
Comment (usamoi) [reply crates/base/src/global/svecf32_cos.rs:22]: They are different.

File: crates/base/src/index.rs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    pub dims: u16,
+    #[serde(rename = "distance")]
+    pub d: DistanceKind,
+    #[serde(rename = "vector")]
+    pub v: VectorKind,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Validate)]
+#[serde(deny_unknown_fields)]
+#[validate(schema(function = "Self::validate_0"))]
```
[the remark this answers is not in the document]
Comment (usamoi) [reply crates/base/src/index.rs:54]: No, as far as I know.

File: src/datatype/binary_svecf32.rs
Comment (usamoi) [reply src/datatype/binary_svecf32.rs:1]: Good idea. We can do it in future.

**Review by cutecutecat — APPROVED** [review]

**Pull request merged**